src/Type.h

```diff
@@ -90,25 +93,25 @@ struct halide_handle_cplusplus_type {
 
     /// Qualifiers and indirections on type. Allows up to 8 levels. 0 is innermost.
     struct CPPTypeModifiers {
-        uint8_t data[8];
-      CPPTypeModifiers(const std::vector<uint8_t> &vals) {
-            user_assert(vals.size() <= sizeof(data)) << "Too many levels of indirection in handle type " << vals.size() << " where " << sizeof(data) << " are allowed\n";
-            std::copy(vals.begin(), vals.end(), &data[0]);
-            std::fill(&data[vals.size()], &data[sizeof(data)], 0);
+        std::array<uint8_t, 8> data;
+        CPPTypeModifiers(const std::vector<uint8_t> &vals) {
+            user_assert(vals.size() <= data.size()) << "Too many levels of indirection in handle type " << vals.size() << " where " << sizeof(data) << " are allowed\n";
+            std::copy(vals.begin(), vals.end(), data.begin());
+            std::fill(data.begin() + vals.size(), data.end(), 0);
         }
         const uint8_t &operator[](size_t index) const { return data[index]; }
         bool operator==(const CPPTypeModifiers &rhs) const {
-            return std::equal(&data[0], &data[sizeof(data)], &rhs.data[0]);
+            return std::equal(data.begin(), data.end(), rhs.data.begin());
         }
-        const uint8_t *begin() const { return &data[0]; }
-        const uint8_t *end() const { return &data[sizeof(data)]; }
+        std::array<uint8_t, 8>::const_iterator begin() const { return data.begin(); }
```

Comment:
Yeah except if it is a const iterator, it should be called const_iterator...

-Z-


On Fri, Jul 29, 2016 at 1:54 PM, Dillon Sharlet <notifications@github.com>
wrote:

> In src/Type.h
> <https://github.com/halide/Halide/pull/1413#discussion_r72858865>:
>
> >          }
> > -        const uint8_t *begin() const { return &data[0]; }
> > -        const uint8_t *end() const { return &data[sizeof(data)]; }
> > +        std::array<uint8_t, 8>::const_iterator begin() const { return data.begin(); }
>
> I think that if we are returning iterators, particularly from methods
> named 'begin' and 'end', that we should do:
>
> typedef std::array<uint8_t, 8>::const_iterator iterator;
> iterator begin() const { ... }
> iterator end() const { ... }
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/halide/Halide/pull/1413/files/929b027f16a1531d221d77847dc3555715ab0cad#r72858865>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABbqFFU-kD9apzCPZ9g-boxMKKmyu7u8ks5qamiXgaJpZM4JXvRH>
> .
>